src/main/java/jeo/parser/Parser.java

```diff
@@ -0,0 +1,182 @@
+ * @author Goh Jun How
+ * @version 0.1
+ */
+public class Parser {
+
+    /**
+     * Accepts a string that represents the trimmed raw string input to be parsed as command and action.
+     * @param s String representing the trimmed raw string input of current line.
+     * @return HashMap linking each essential aspect of the input to its corresponding substring.
+     * @throws JeoException for custom errors.
+     */
+    public static HashMap<String, String> parseString(String s) throws JeoException {
+        HashMap<String, String> hm = new HashMap<>();
+        String str = s.trim();
+        if (str.contains("\\")) {
+            throw new JeoException("Backslash character \"\\\" not allowed");
+        }
+        if (str.equalsIgnoreCase("bye")) {
+            hm.put("command", "bye");
+        } else if (str.equalsIgnoreCase("list")) {
+            hm.put("command", "list");
+        } else if (str.toLowerCase().startsWith("todo")) {
+            hm.put("command", "todo");
+            str = str.substring(4).trim();
+            hm.put("description", str);
+            if (str.isEmpty()) {
+                throw new JeoException("Please enter a task description.");
+            }
+        } else if (str.toLowerCase().startsWith("deadline")) {
+            hm.put("command", "deadline");
+            str = str.substring(8).trim();
+            if (!str.toLowerCase().contains("/by")) {
+                throw new JeoException(str.isEmpty()
+                        ? "Please enter a task description."
+                        : "Please follow the format: deadline <description> /by <yyyy-MM-dd HH:mm>");
+            }
+            String[] arr = parseSubstringTasks(str, "deadline");
+            String desc = arr[0];
+            String by = arr[1];
+            if (desc.isEmpty()) {
+                throw new JeoException("Please enter a task description.");
+            }
+            if (by.isEmpty()) {
+                throw new JeoException("Please enter a date-time in the format "
+                        + "\"yyyy-MM-dd HH:mm\" after \"/by\".");
+            }
+            hm.put("description", desc);
+            hm.put("by", by);
+        } else if (str.toLowerCase().startsWith("event")) {
+            hm.put("command", "event");
+            str = str.substring(5).trim();
+            if (!str.toLowerCase().contains("/from") || !str.toLowerCase().contains("/to")) {
+                throw new JeoException(str.isEmpty()
+                        ? "Please enter a task description."
+                        : "Please follow the format: event <description> "
+                        + "/from <yyyy-MM-dd HH:mm> /to <yyyy-MM-dd HH:mm>");
+            }
+            String[] arr = parseSubstringTasks(str, "event1");
+            String desc = arr[0];
+            str = arr[1];
+            arr = parseSubstringTasks(str, "event2");
+            String from = arr[0];
+            String to = arr[1];
```

Comment:
I think u can consider replacing the if-else structure with switch case for better code readability.